whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java

```diff
@@ -3,96 +3,134 @@
-                LOGGER.debug("Sync Whois changes to snapshot here (not implemented)");
-            }
-            final PublishableSnapshotFile snapshotFile = new PublishableSnapshotFile(sourceVersion);
-            final String fileName = NrtmFileUtil.newFileName(snapshotFile);
-            Stopwatch stopwatch = Stopwatch.createStarted();
-            try (final OutputStream out = nrtmFileStore.getFileOutputStream(snapshotFile.getSessionID(), fileName)) {
-                snapshotFileSerializer.writeSnapshotAsJson(snapshotFile, out);
-            }
-            LOGGER.info("Wrote JSON for {} in {}", snapshotFile.getSourceModel(), stopwatch);
+            LOGGER.info("Creating snapshot for {}", sourceVersion.getSource().getName());
+            final PublishableNrtmFile snapshotFile = new PublishableNrtmFile(sourceVersion);
+            final String fileName = NrtmFileUtil.newGzFileName(snapshotFile);
+            snapshotFile.setFileName(fileName);
+            publishedFiles.add(snapshotFile);
+            final LinkedBlockingQueue<RpslObjectData> queue = new LinkedBlockingQueue<>(QUEUE_CAPACITY);
+            queueMap.put(snapshotFile.getSource().getName(), queue);
+            final RunnableFileGenerator runner = new RunnableFileGenerator(nrtmFileService, snapshotFileRepository, snapshotFileSerializer, snapshotFile, queue);
+            final Thread queueReader = new Thread(runner);
+            queueReader.start();
+            queueReaders.add(queueReader);
+        }
+        new Thread(rpslObjectEnqueuer.getRunner(state, queueMap)).start();
+        for (final Thread queueReader : queueReaders) {
             try {
-                stopwatch = Stopwatch.createStarted();
-                final String sha256hex = DigestUtils.sha256Hex(nrtmFileStore.getFileInputStream(snapshotFile.getSessionID(), fileName));
-                snapshotFile.setFileName(fileName);
-                snapshotFile.setHash(sha256hex);
-                snapshotFileRepository.insert(snapshotFile);
-                LOGGER.info("Calculated hash for {} in {}", snapshotFile.getSourceModel(), stopwatch);
-                snapshotFiles.add(snapshotFile);
-            } catch (final IOException e) {
-                LOGGER.error("Exception thrown when calculating hash of snapshot file", e);
-                throw e;
+                queueReader.join();
+            } catch (final InterruptedException e) {
+                LOGGER.info("Queue reader (JSON file writer) interrupted", e);
+                Thread.currentThread().interrupt();
             }
         }
-        return snapshotFiles;
+        LOGGER.info("Generation complete {}", stopwatchRoot);
+        return publishedFiles;
     }
 
     public Optional<SnapshotFile> getLastSnapshot(final NrtmSourceModel source) {
         return snapshotFileRepository.getLastSnapshot(source);
     }
 
+    private record RunnableFileGenerator(NrtmFileService nrtmFileService, SnapshotFileRepository snapshotFileRepository,
```

Comment:
these parameters should be final, right?

Reply:
no -- a record is immutable by default so 'final' is implied.